File: core/src/main/java/javabot/JavabotModule.java
package javabot;

import com.antwerkz.sofia.Sofia;
import com.google.inject.AbstractModule;
import com.google.inject.Provider;
import com.google.inject.Provides;
import com.mongodb.MongoClient;
import com.mongodb.MongoClientOptions;
import com.mongodb.ServerAddress;
import com.mongodb.WriteConcern;
import javabot.dao.ChannelDao;
import javabot.dao.ConfigDao;
import javabot.dao.util.LocalDateTimeConverter;
import javabot.javadoc.JavadocClass;
import javabot.model.Channel;
import javabot.model.Config;
import javabot.model.Factoid;
import javabot.web.JavabotApplication;
import net.swisstech.bitly.BitlyClient;
import org.aeonbits.owner.Config.Key;
import org.aeonbits.owner.ConfigFactory;
import org.mongodb.morphia.Datastore;
import org.mongodb.morphia.Morphia;
import org.pircbotx.Configuration.Builder;
import org.pircbotx.PircBotX;
import org.pircbotx.cap.SASLCapHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Singleton;
import java.io.IOException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class JavabotModule extends AbstractModule {

    private static final Logger LOG = LoggerFactory.getLogger(JavabotModule.class);

    private MongoClient mongoClient;

    private Morphia morphia;

    private Datastore datastore;
    private Provider<BotListener> botListenerProvider;

    protected Provider<ChannelDao> channelDaoProvider;
    protected Provider<ConfigDao> configDaoProvider;

    @Override
    protected void configure() {
        configDaoProvider = binder().getProvider(ConfigDao.class);
        channelDaoProvider = binder().getProvider(ChannelDao.class);
        botListenerProvider = binder().getProvider(BotListener.class);
    }

    @Provides
    @Singleton
    public Datastore datastore() throws IOException {
        if (datastore == null) {
            datastore = getMorphia().createDatastore(getMongoClient(), javabotConfig().databaseName());
            datastore.setDefaultWriteConcern(WriteConcern.SAFE);
            try {
                datastore.ensureIndexes();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return datastore;
    }

    @Provides
    @Singleton
    public Morphia getMorphia() {
        if (morphia == null) {
            morphia = new Morphia();
            morphia.mapPackage(JavadocClass.class.getPackage().getName());
            morphia.mapPackage(Factoid.class.getPackage().getName());
            morphia.getMapper().getConverters().addConverter(LocalDateTimeConverter.class);
        }
        return morphia;
    }

    @Provides
    @Singleton
    public MongoClient getMongoClient() throws IOException {
        if (mongoClient == null) {
            try {
                mongoClient = new MongoClient(new ServerAddress(javabotConfig().databaseHost(), javabotConfig().databasePort()),
                                              MongoClientOptions.builder()
                                                                .connectTimeout(2000)
                                                                .build());
            } catch (RuntimeException e) {
                e.printStackTrace();
                throw new RuntimeException(e.getMessage(), e);
            }
        }
        return mongoClient;
    }

    @Provides
    @Singleton
    public BitlyClient bitlyClient() throws IOException {
        return javabotConfig().bitlyToken() != null ? new BitlyClient(javabotConfig().bitlyToken()) : null;
    }

    @Provides
    @Singleton
    protected PircBotX createIrcBot() {
        Config config = configDaoProvider.get().get();
        String nick = config.getNick();
        Builder<PircBotX> builder = new Builder<>()
                                        .setName(nick)
                                        .setLogin(nick)
                                        .setAutoNickChange(false)
                                        .setCapEnabled(false)
                                        .addListener(getBotListener())
                                        .setServerHostname(config.getServer())
                                        .setServerPort(config.getPort())
                                        .addCapHandler(new SASLCapHandler(nick, config.getPassword()));
/*
        for (Channel channel : channelDaoProvider.get().getChannels()) {
            LOG.info("Adding {} as an autojoined channel", channel.getName());
            if (channel.getKey() == null) {
                builder.addAutoJoinChannel(channel.getName());
            } else {
                builder.addAutoJoinChannel(channel.getName(), channel.getKey());
            }
        }
*/

        return new PircBotX(builder.buildConfiguration());
    }

    @Provides
    @Singleton
    public JavabotConfig javabotConfig() throws IOException {
        return validate(ConfigFactory.create(JavabotConfig.class, new HashMap<>(), System.getProperties(), System.getenv()));
    }

    @SuppressWarnings("unchecked")
    protected JavabotConfig validate(final JavabotConfig config) {
        Class<? extends JavabotConfig> configClass = (Class<? extends JavabotConfig>) config.getClass().getInterfaces()[0];
        Method[] methods = configClass.getDeclaredMethods();
        List<String> missingKeys = new ArrayList<>();
        for (Method method : methods) {
            try {
                Key annotation = method.getDeclaredAnnotation(Key.class);
                if (annotation != null && method.getParameterCount() == 0
                    && !method.getReturnType().equals(Void.class)
                    && method.invoke(config) == null) {
                    missingKeys.add(annotation.value());
                }
            } catch (ReflectiveOperationException e) {
                throw new RuntimeException(e.getMessage(), e);
            }
        }
        if(!missingKeys.isEmpty()) {
            throw new RuntimeException(Sofia.configurationMissingProperties(missingKeys));
        }
        return config;
    }

    public BotListener getBotListener() {
        return botListenerProvider.get();
    }
}

File: src/main/kotlin/javabot/JavabotModule.kt
package javabot

import com.antwerkz.sofia.Sofia
import com.google.common.base.CharMatcher
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import dev.morphia.Datastore
import dev.morphia.Morphia
import dev.morphia.config.MorphiaConfig
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.inject.Produces
import jakarta.inject.Inject
import jakarta.inject.Singleton
import javabot.dao.ConfigDao
import javabot.model.Factoid
import javabot.model.javadoc.JavadocClass
import javax.net.ssl.SSLSocketFactory
import net.thauvin.erik.bitly.Bitly
import org.aeonbits.owner.Config.Key
import org.aeonbits.owner.ConfigFactory
import org.pircbotx.Configuration.Builder
import org.pircbotx.PircBotX
import org.pircbotx.cap.SASLCapHandler

@ApplicationScoped
open class JavabotModule @Inject constructor(val configDao: ConfigDao, val ircAdapter: IrcAdapter) {
    open val mongoClient: MongoClient by lazy {
        MongoClients.create(MongoClientSettings.builder().build())
    }

    private var config: JavabotConfig? = null
    //    lateinit var ircAdapterProvider: Instance<out IrcAdapter>
    //    lateinit var channelDaoProvider: Instance<ChannelDao>
    //    lateinit var configDaoProvider: Instance<ConfigDao>

    open fun configure() {
        //        configDaoProvider = binder().getProvider(ConfigDao::class.java)
        //        channelDaoProvider = binder().getProvider(ChannelDao::class.java)
        //        ircAdapterProvider = binder().getProvider(IrcAdapter::class.java)
        //        install(FactoryModuleBuilder().build(ViewFactory::class.java))
    }

    open fun client(): MongoClient {
        return mongoClient
    }

    @Produces
    @Singleton
    fun datastore(): Datastore {
        val databaseName: String = javabotConfig().databaseName()
        val datastore =
            Morphia.createDatastore(
                client(),
                MorphiaConfig.load()
                    .applyIndexes(true)
                    .autoImportModels(true)
                    .database(databaseName)
                    .enablePolymorphicQueries(true)
                    .packages(
                        listOf(
                            JavadocClass::class.java.packageName,
                            Factoid::class.java.packageName
                        )
                    )
            )

        return datastore
    }

    @Produces
    @Singleton
    protected open fun createIrcBot(): PircBotX {
        val config = configDao.get()

        return PircBotX(
            Builder()
                .setName(getBotNick())
                .setLogin(getBotNick())
                .setAutoNickChange(false)
                .setCapEnabled(false)
                .addListener(ircAdapter)
                .addServer(config.server, config.port)
                .addCapHandler(SASLCapHandler(getBotNick(), config.password))
                .setSocketFactory(SSLSocketFactory.getDefault())
                .buildConfiguration()
        )

        /*
                return object: PircBotX(builder.buildConfiguration()) {
                    override fun sendRawLineToServer(line: String) {

                        var line = line
                        if (line.length > configuration.maxLineLength - 2) line = line.substring(0, configuration.maxLineLength - 2)
                        println("raw line: $line")
                        outputWriter.write(line + "\r\n")
                        outputWriter.flush()
                        val lineParts = tokenizeLine(line)
                        getConfiguration().getListenerManager<ListenerManager>().onEvent(OutputEvent(this, line, lineParts))
                    }
                }
        */
    }

    protected open fun getBotNick(): String {
        return configDao.get().nick
    }

    @Produces
    @Singleton
    fun javabotConfig(): JavabotConfig {
        if (config == null) {
            config =
                ConfigFactory.create(
                    JavabotConfig::class.java,
                    loadConfigProperties(),
                    System.getProperties(),
                    System.getenv()
                )
            validate(config!!)
        }
        return config!!
    }

    @Produces
    @Singleton
    fun bitly(): Bitly? {
        val bitlyToken = javabotConfig().bitlyToken()
        return if (bitlyToken != "") Bitly(bitlyToken) else null
    }

    protected open fun loadConfigProperties(): HashMap<Any, Any> = HashMap()

    protected fun validate(config: JavabotConfig): JavabotConfig {
        @Suppress("UNCHECKED_CAST")
        val configClass = config.javaClass.interfaces[0] as Class<JavabotConfig>
        val methods = configClass.declaredMethods
        val missingKeys = ArrayList<String>()
        for (method in methods) {
            try {
                val annotation = method.getDeclaredAnnotation(Key::class.java)
                if (
                    annotation != null &&
                        method.parameterCount == 0 &&
                        method.returnType != Void::class.java &&
                        method.invoke(config) == null
                ) {
                    missingKeys.add(annotation.value)
                }
            } catch (e: ReflectiveOperationException) {
                throw RuntimeException(e.message, e)
            }
        }
        if (missingKeys.isNotEmpty()) {
            throw RuntimeException(Sofia.configurationMissingProperties(missingKeys))
        }
        return config
    }
}

fun tokenizeLine(input: String?): MutableList<String> {
    val stringParts: MutableList<String> = java.util.ArrayList()
    if (input.isNullOrEmpty()) return stringParts
    // Heavily optimized string split by space with all characters after :
    // added as a single entry. Under benchmarks, this is faster than
    // StringTokenizer, String.split, toCharArray, and charAt
    val trimmedInput: String = CharMatcher.whitespace().trimFrom(input)
    var pos = 0
    var end: Int
    while (trimmedInput.indexOf(' ', pos).also { end = it } >= 0) {
        stringParts.add(trimmedInput.substring(pos, end))
        pos = end + 1
        if (trimmedInput[pos] == ':') {
            stringParts.add(trimmedInput.substring(pos + 1))
            return stringParts
        }
    }
    // No more spaces, add last part of line
    stringParts.add(trimmedInput.substring(pos))
    return stringParts
}
